### PR TITLE
Adding PCA and Tfac

### PR DESCRIPTION
**Pull request by andrewram4287**
Adding decomposition of PCA and Tfac for single cell data. Add to R2X plotting as well. Using tensorpack version of tensor decomposition and PCA

### CODE REVIEW

**Review by aarmey — COMMENTED** [review]
What is the purpose of using tensorpack? I thought plotR2X there would only be for CP...

**Comment by andrewram4287** [comment]
> What is the purpose of using tensorpack? I thought plotR2X there would only be for CP...

I wanted to compare Pf2's R2X to other methods, at least for the synthetic data. This is from my proposal

To evaluate Pf2’s ability to reduce the data dimensionality and determine the optimal rank to explain the variance in the data, we will measure Pf2’s reconstruction using R2X, which defines the percentage of variance explained based on the number of components. To compare Pf2, we will be performing Cp TD, non-negative matrix factorization (NNMF), and PCA on the same synthetic datasets and using the same number of components as seen in Figure 3D. Pf2 should outperform the other methods by explaining more variance with fewer components.

**Comment by aarmey** [comment]
Aha—I should have caught it there, but there's really no way to meaningfully use CP here. I think you can just have PARAFAC2, PCA, and NNMF.

**Comment by andrewram4287** [comment]
> Aha—I should have caught it there, but there's really no way to meaningfully use CP here. I think you can just have PARAFAC2, PCA, and NNMF.

Ok sounds good! For NNMF, how should I manipulate the data to be positive since it can't accept negative values? Or should I just change all the synthetic data to be positive?

**Comment by aarmey** [comment]
I think it makes sense to only apply NNMF where non-negativity would be a natural choice (e.g., scRNAseq).

By the way, we will want to express this as R2X vs. the degrees of freedom of the decomposition. The number of components doesn't mean the same thing between each type of model. I believe the DoF are a bit tricky to calculate for PARAFAC2 because of the projection matrices. We can always setup the code for # of components and then figure out DoF later.

File: sccp/decomposition.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+                )
+
+            if len(tensor) > 3:
+                projs = np.reshape(projs, (-1, tensor.shape[-2], i+1))
+                factors = [khatri_rao(factors[:-2]), factors[-2], factors[-1]]
+
+            pf2_error[i] = 1 - _parafac2_reconstruction_error(
+                tensor, (weights, factors, projs))/ np.linalg.norm(tensor) ** 2
+            
+            decomp = Decomposition(tensor, max_rr=i+1)
```
Comment (aarmey) [inline sccp/decomposition.py:34]: This gives you the reconstruction error with PCA for all ranks, so this can be outside the loop.

**Review by aarmey — APPROVED** [review]
One change, then this is good to merge.

**Pull request merged**